doc/command_guidelines.md

```diff
@@ -121,6 +130,252 @@ For commands that don't conform to one of the above-listed standard command patt
+        return FooRule(**kwargs)
+```
+
+The usage pattern of this implementation to create two rules would be:
+```
+az foo create -g MyRG -n myfoo --action allow --values test prod --rule name=rule1 metric=name operation=equals value=test --rule name=rule2 metric=age operation=lessThan value=100
+```
+
+### Overcoming Service Limitations
+
+There are times when service limitations can interfere with the recommended approach of handling complex collections using subcommand groups. In our example, the service might fail if the `foo_create` command supplied an empty rule collection. This is something we can work around in the CLI using the `--defer` caching mechanism. The following changes should be made:
+
+In `commands.py`, add the `supports_local_cache=True` kwarg to the command registration:
+```Python
+with self.command_group('foo', foo_sdk) as g:
+    g.custom_command('create', 'create_foo', supports_local_cache=True)
+    ...
+
+with self.command_group('foo rule', foo_sdk) as g:
+    g.custom_command('create', 'create_foo_rule', supports_local_cache=True)
+    ... 
+```
+
+Update the custom commands to use the `cached_get` and `cached_put` helpers:
+```Python
+def create_foo(cmd, client, resource_group_name, foo_name, policy_action, policy_values):
+    from azure.cli.core.commands import cached_get, cached_put
+    Foo, Policy = cmd.get_models('Foo', 'Policy')
+    foo_obj = Foo(
+        rules=[],
+        policy=Policy(
+            action=policy_action,
+            values=policy_values
+        )
+    )
+    # cache the payload if --defer used or send to Azure
+    return cached_put(cmd, client.create_or_update, resource_group_name, foo_name, foo_obj)
+```
+
+The corresponding rule create looks like:
+```Python
+def create_foo_rule(cmd, client, resource_group_name, foo_name, rule_name, metric, operation, value):
+    from azure.cli.core.commands import cached_put, upsert_to_colleciton, get_property
+    FooRule = cmd.get_models('FooRule')
+    # retrieves the object from the cache. On a miss, retrieves from Azure
+    foo_obj = cached_get(cmd, client.get, resource_group_name, foo_name)
+    new_rule = FooRule(
+        name=rule_name,
+        metric=metric,
+        operation=operation,
+        value=value
+    )
+    # add the new child to the parent collection
+    upsert_to_collection(foo_obj, 'rules', new_rule, 'name')
+    # re-cache the parent with --defer or send to Azure
+    foo_obj = cached_put(cmd, client.create_or_update, name, resource_group_name, foo_obj).result()
+    # return the child object
+    return get_property(foo_obj.rules, rule_name)
+```
+
+The usage pattern of this implementation to create two rules would be:
+```
+az foo create -g MyRG -n myfoo --action allow --values test prod --defer
+az foo rule create -g MyRG --foo-name myfoo -n rule1 --metric name --operations equals --value test --defer
+az foo rule create -g MyRG --foo-name myfoo -n rule2 --metric age --operations lessThan --value 100
+```
+
```

Comment:
Should we had a minor comment that cache objects can be managed with az cache. And their ttl can be set via az configure?

Reply:
I see no reason to. It's irrelevant from the authoring pattern, which is what this is about.

Reply:
Makes sense. I just saw too that --defer does document az cache. So this sounds good to me!

Reply:
I added a reference to it.